File: KthSmallestEleBT.py
# Definition for a binary tree node.
from typing import Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        self.count, self.res = 0, 0
    
        def dfs(root, k):
            if not root or self.count >= k:
                return
            dfs(root.left, k)

            self.count += 1
            if self.count == k:
                self.res = root.val
                return 
            dfs(root.right, k)
        dfs(root, k)
        return self.res

node2 = TreeNode(2, None, None)
node4 = TreeNode(4, None, None)
node1 = TreeNode(1, None, node2)
node3 = TreeNode(3, node1, node4)


print(Solution().kthSmallest(node3, 1))

File: KthSmallestEleBT.java
class KthSmallestEleBT {

    int count = 0;
    int res = 0;

    public static void main(String[] args) {
        
        TreeNode node2 = new TreeNode(2, null, null);
        TreeNode node4 = new TreeNode(4, null, null);
        TreeNode node1 = new TreeNode(1, null, node2);
        TreeNode node3 = new TreeNode(3, node1, node4);

        
        KthSmallestEleBT ans = new KthSmallestEleBT();
        System.out.println(ans.kthSmallest(node3, 1));


    }

    public int kthSmallest(TreeNode root, int k) {
        helper(root, k);
        return res;
    }

    private void helper(TreeNode root, int k) {
        if (root == null || count >= k) {
            return ;
        }

        helper(root.left, k);
        
        count++;
        if (count == k) {
            res = root.val;
            return;
        }

        helper(root.right, k);
    }
}

public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

